realm/realm-library/src/main/java/io/realm/Realm.java

```diff
@@ -1092,15 +1091,23 @@ public void run() {
                     if (handler != null
                             && !Thread.currentThread().isInterrupted()
                             && handler.getLooper().getThread().isAlive()) {
-                        if (onSuccess != null && transactionCommitted) {
+
+                        if (transactionCommitted) {
+                            // This will be treated like a special REALM_CHANGED event
                             handler.post(new Runnable() {
                                 @Override
                                 public void run() {
-                                    onSuccess.onSuccess();
+                                    handlerController.handleAsyncTransactionCompleted(onSuccess != null ? new Runnable() {
+                                        @Override
+                                        public void run() {
+                                            onSuccess.onSuccess();
+                                        }
+                                    } : null);
```

Comment:
Using the REALM_CHANGED if `onSuccess` is null will actually bloat this code quite a bit as we then need to send a empty message to the handler instead. By doing it this way we hide that complexity in `handleAsyncTransactionCompleted`